### ISSUE TITLE
Expose Type from Prelude

### ISSUE THREAD

**Issue by noughtmare** [haskell/core-libraries-committee#193]
### Motivation

Now that `PolyKinds` is enabled by default as part of `GHC2021`, it has become much more common that you have to specify the kind `Type` explicitly to avoid ambiguity. (TODO: add example) 

Currently, to do that idiomatically you need to import `Type` from `Data.Kind`, e.g.:

```haskell
import Data.Kind (Type)
```

It is possible to avoid additional imports by using `*` because `StarIsType` is still enabled by default too, but that is likely to be removed in the future. So I think the status quo is nudging people in the wrong direction.

### Changes

The change should be quite simple, but I will work it out later.

### Impact

This is a breaking change because people might have already used the name `Type` for something else and I expect it to cause a non-trivial amount of breakage, unfortunately. I will try to do a more detailed impact assessment and write a migration guide, but I'd love it if someone could help me with that.

### WIP

I will update this with more details and examples, but I wanted to write this down somewhere.

**Comment by phadej** [comment]
> This is a breaking change because people might have already used the name Type for something else

- `ghc` itself: [`Type`](https://hackage.haskell.org/package/ghc-9.6.1/docs/GHC-Core-Type.html#t:Type) is used everywhere. 
- `template-haskell`: [`Type`](https://hackage.haskell.org/package/template-haskell-2.20.0.0/docs/Language-Haskell-TH.html#t:Type)
  Template Haskell is probably more common than GHC2021 usage.
  
So I expect impact to be huge.

**Comment by googleson78** [comment]
> I will try to do a more detailed impact assessment

I can try to help with this, however, last time I tried, I was [blocked on using clc-stackage](https://github.com/Bodigrim/clc-stackage/issues/14), so I need to figure that out before I can help.

**Comment by noughtmare** [comment]
> Template Haskell is probably more common than GHC2021 usage.

I'd say, if we want the language to keep living, GHC2021 (or some other new standard) usage must eventually surpass Haskell2010 usage.

Maybe we should rip off the band-aid as fast as possible, but we should make sure to have a very good migration strategy. I think it would be feasible and desireable to use a tool like `retrie` to automate this change.

**Comment by tomjaguarpaw** [comment]
I also think we try our hardest to expose `Type` from the `Prelude` given that `*` is going to go away at some point. Doing so will cause vast breakage which is very unfortunate and suggestions for how to mitigate the breakage.

**Comment by TeofilC** [comment]
This search: https://hackage-search.serokell.io/?q=%5E%28data%7Cnewtype%7Ctype%29+Type%28%5Cs%7C%3D%29
suggests that there are approximately 337 packages that define something called `Type`, though some of these are probably false positives.

**Comment by noughtmare** [comment]
> though some of these are probably false positives.

And there may be false negatives. For example, it would miss data and type families, promoted data kinds, and indentation.

**Comment by parsonsmatt** [comment]
I agree that the `Type` name is too much of a conflict for `Prelude` export. It *is* annoying already to have `Data.Kind` and `template-haskell` in scope at the same time.

It does seem like the concept should be available pervasively, much like `a :: *` requires no imports or extensions. Possibly a new name may be suitable for export from `Prelude`.

Adding things to `Prelude` is a heavyweight tool, especially when custom preludes are so easy in Haskell.

**Comment by tomjaguarpaw** [comment]
> Possibly a new name may be suitable for export from Prelude.

Do you mean a type synonym? Did you have a suggestion?

**Comment by googleson78** [comment]
> But it doesn't need to have that name, does it?

In that case it seems doubtful to me that it actually helps with migrating away from `*` (in the "no imports" case), as I doubt many people will prefer to write `(HaskellType -> HaskellType) -> HaskellType` as opposed to `(* -> *) -> *`. Maybe I'm just misunderstanding the point of this proposal though.

**Comment by mixphix** [comment]
`type Hask = Data.Kind.Type` would line up with much existing literature.

**Comment by tomjaguarpaw** [comment]
> In that case it seems doubtful to me that it actually helps with migrating

Fair enough. Then I think the Haskell community needs to hold a Difficult Conversation™ regarding what we're going to do about the transition from `*` to `Type`.

**Comment by Bodigrim** [comment]
Looking at the original proposal [Remove the * kind syntax](https://ghc-proposals.readthedocs.io/en/latest/proposals/0143-remove-star-kind.html#breakage-estimation): oh blessed times when the world was young and breaking 25% of Hackage for the sake of Dependent Haskell quibbles was deemed acceptable! But I digress.

----

I'm with @mixphix on this: `import Data.Kind (Type)` is sufficiently good. There is more than enough breakage associated with the eventual removal of `*`, no need to deepen the wound.

> It is my current understanding that `PolyKinds` already causes all parameterised data type definitions to get polymorphic kinds inferred, so you're not really going to escape type-level stuff in most modules.

This is not my experience. I maintain a large codebase with `-Werror -Wall -Wcompat` and `GHC2021`, and I still retain enough fingers on my hands to count all occurrences of `import Data.Kind` there. The codebase will be absolutely broken by `Type` in `Prelude` however, because of name clashes.

**Comment by Bodigrim** [comment]
> But it doesn't need to have that name, does it? How about a type synonym, say `type HaskellType = Data.Kind.Type`?

`Type` is already a [`type`](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Kind.html#t:Type) itself. One option is to export `TYPE` and `LiftedRep` from `Prelude`, which seems weird, but unlikely to generate many name clashes.

**Comment by Ericson2314** [comment]
> `type Hask = Data.Kind.Type` would line up with much existing literature.

I would expect that to be `(->)` not `Type`, given how `Control.Category.Category` works.

**Comment by mixphix** [comment]
Categories can be named after their morphisms (like `Mat_k`), but usually they're called by their objects (`Set`, `Top`, `Grp`, `Ring`, `Cat`, `Vect_k`, ...). But indeed [`Hask` is not a category](https://stackoverflow.com/a/48490015), and `Control.Category.Category` is only the barest of approximations to actual category theory. So the synonym would be mainly mnemonic.

**Comment by noughtmare** [comment]
One more reason why `Type` is becoming more important: the warning ` -Wmissing-kind-signatures` implies that it is good practice to annotate every data type declaration (just like how it is good practice to give a type signature to every top level function). Most will require `Type`.

**Comment by Bodigrim** [comment]
`-Wmissing-kind-signatures` is not enabled even by `-Wall`.

**Comment by Bodigrim** [comment]
@noughtmare if you are still interested in the proposal, one has to make some progress towards impact assessment.

**Comment by noughtmare** [comment]
@Bodigrim I'd like to withdraw this proposal. I still think it would be very useful for certain use-cases which are going to become more common as Haskell moves more towards dependent types, however I see no way to get there without a massive effort. Perhaps if one day the issue becomes more pressing or if we get better refactoring tools, I will revisit this. 

Thank you all that participated in this thread for the useful feedback and discussion.

**Comment by noughtmare** [comment]
I don't want to reopen this proposal, but for posterity I have an example to add to my claim:

> It is my current understanding that PolyKinds already causes all parameterised data type definitions to get polymorphic kinds inferred, so you're not really going to escape type-level stuff in most modules. I do realise this needs to be substantiated with a good example which is currently my highest priority for this proposal.

Here's a concrete example of what I meant: https://discourse.haskell.org/t/issue-with-overlapping-instances/11467?u=jaror

**Comment by noughtmare** [comment]
More evidence:

https://discourse.haskell.org/t/a-very-niche-footgun-in-ghc-generics/11779?u=jaror
